### PR TITLE
fixes #980 - apoc.load.xls Cannot convert java.util.Date to AnyValue

### LINKED ISSUE #980: apoc.load.xls Cannot convert java.util.Date to AnyValue
Hi everyone,

I could not find anything about the error i'm getting trying to use apoc.load.xls. In my xlsx file I have a column that contains value displayed as Date and when using this command 

```
CALL apoc.load.xls('file:///C:/file.xlsx','sheet',{mapping:{Date:{type:'String'}}}) YIELD lineNo, list, map
```

I'm getting this error :
```
Error
Neo.DatabaseError.General.UnknownError
Neo.DatabaseError.General.UnknownError: Cannot convert java.util.Date to AnyValue
```

**OS :** Windows 7 Pro
**Neo4j :** community-3.4.9
**Plugins :** poi-3.17.jar, commons-collections4-4.1.jar, poi-ooxml-3.17.jar, poi-ooxml-schemas-3.17.jar, xmlbeans-2.6.0.jar, curvesapi-1.04.jar

I've seen similar error but was with import from mongodb https://github.com/neo4j-contrib/neo4j-apoc-procedures/issues/355 
Thanks in advance !

### PR DESCRIPTION
**Pull request by AngeloBusato**
Fixes #980 
 
apoc.load.xls Cannot convert java.util.Date to AnyValue

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

  - Add config `dateFormat` and `dateParse` to the mapping conf e.g {mapping:{'<sheet>':{type:'<type>', dateFormat: '<format>', dateParse: [<formats>]}}}
  - Now is possible to convert the String to Date or Date to String

### CODE REVIEW

File: src/main/java/apoc/temporal/TemporalProcedures.java [hunk 1]
```diff
@@ -13,65 +14,68 @@
...
-        ISO_DATE_FORMAT.put("week_date", DateTimeFormatter.ofPattern("xxxx-'W'ww-e"));
-        ISO_DATE_FORMAT.put("week_date_time", DateTimeFormatter.ofPattern("xxxx-'W'ww-e'T'HH:mm:ss.SSSZZ"));
-        ISO_DATE_FORMAT.put("week_date_time_no_millis", DateTimeFormatter.ofPattern("xxxx-'W'ww-e'T'HH:mm:ssZZ"));
-        ISO_DATE_FORMAT.put("weekyear", DateTimeFormatter.ofPattern("xxxx"));
-        ISO_DATE_FORMAT.put("weekyear_week", DateTimeFormatter.ofPattern("xxxx-'W'ww"));
-        ISO_DATE_FORMAT.put("weekyear_week_day", DateTimeFormatter.ofPattern("xxxx-'W'ww-e"));
-        ISO_DATE_FORMAT.put("year", DateTimeFormatter.ofPattern("yyyy"));
-        ISO_DATE_FORMAT.put("year_month", DateTimeFormatter.ofPattern("yyyy-MM"));
-        ISO_DATE_FORMAT.put("year_month_day", DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+    static {
```
Comment (jexp) [inline src/main/java/apoc/temporal/TemporalProcedures.java]: can we move this to a separate class and use Util.map() and/or a helper method to make this setup more readable.

File: docs/loadcsv.adoc [hunk 2]
```diff
@@ -250,8 +251,27 @@ The usage of `apoc.load.xls` is similar to apoc.load.csv with the main differenc
...
+| dateParse: [<formats>] | Convert the String into Date (Array of strings are allowed)
+|===
+
+==== Note
+
+In dateParse the first format matched return the date formatted, otherwise it will return an error
 
+In `format` config you can use the pattern describe as the Temporal functions: link:#_temporal[temporal funcionts]
+
+=== Examples for apoc.load.xls
```
Comment (jexp) [inline docs/loadcsv.adoc]: can you add a concrete example for date type(s) and one for dateParse/dateFormat?

File: src/main/java/apoc/util/Util.java [hunk 3]
```diff
@@ -700,4 +705,47 @@ public static boolean classExists(String className) {
         Object value = ApocConfiguration.get(loadType).get(key);
         return Optional.ofNullable(value).map(Object::toString);
     }
+
+    public static String dateFormat(TemporalAccessor value, String format){
+        return getFormat(format).format(value);
+    }
+
+    public static TemporalAccessor dateParse(String value, Class<? extends TemporalAccessor> date, String...formats) {
+        String methodName = "parse";
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: Why do we have to do it like this?
Isn't that on an interface?

File: src/main/java/apoc/load/LoadXls.java [hunk 4]
```diff
@@ -20,7 +24,7 @@
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
-import static apoc.util.Util.cleanUrl;
+import static apoc.util.Util.*;
```
Comment (jexp) [inline src/main/java/apoc/load/LoadXls.java]: please also update the procedure `@Description`

File: src/main/java/apoc/util/Util.java (hunk 3 above)
Comment (conker84) [reply src/main/java/apoc/util/Util.java]: Do you mean that can be invoked in this way?
```java
parseDate(value.toString(), TemporalAccessor.class, dateParse);
```

File: src/main/java/apoc/temporal/TemporalProcedures.java (hunk 1 above)
Comment (conker84) [reply src/main/java/apoc/temporal/TemporalProcedures.java]: Ok

File: src/main/java/apoc/load/LoadXls.java (hunk 4 above)
Comment (conker84) [reply src/main/java/apoc/load/LoadXls.java]: Ok

File: docs/loadcsv.adoc (hunk 2 above)
Comment (conker84) [reply docs/loadcsv.adoc]: Sure

File: src/main/java/apoc/util/Util.java (hunk 3 above)
Comment (jexp) [reply src/main/java/apoc/util/Util.java]: create method-handle and cache them in a static Map<Class,MethodHandle>

**Comment by AngeloBusato** [comment]
@jexp PR Updated

File: src/main/java/apoc/load/LoadXls.java [hunk 5]
```diff
@@ -153,7 +159,8 @@ public Mapping(String name, Map<String, Object> mapping, char arraySep, boolean
             this.arraySep = separator(mapping.getOrDefault("arraySep", arraySep).toString(),DEFAULT_ARRAY_SEP);
             this.type = Meta.Types.from(mapping.getOrDefault("type", "STRING").toString());
             this.arrayPattern = Pattern.compile(String.valueOf(this.arraySep), Pattern.LITERAL);
-
+            this.dateFormat = mapping.getOrDefault("dateFormat", StringUtils.EMPTY).toString();
```
[the remark this answers is not in the document]
Comment (conker84) [reply src/main/java/apoc/load/LoadXls.java]: Just to simplify the if check, so we can exclude a null check and provide only the `isEmpty`

File: src/main/java/apoc/util/DateParseUtil.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        } catch (Throwable throwable) {
+            throwable.printStackTrace();
+        }
+        throw new RuntimeException("Can't format the date with the pattern");
+    }
+
+    private static TemporalAccessor getParse(Class<? extends TemporalAccessor> date, DateTimeFormatter format, String value) throws Throwable {
+        MethodHandle methodHandle;
+
+        if (parseDateMap.containsKey(date)) {
```
Comment (jexp) [inline src/main/java/apoc/util/DateParseUtil.java]: computeIfAbsent would make most sense.

File: src/main/java/apoc/util/DateParseUtil.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import java.time.format.DateTimeParseException;
+import java.time.temporal.TemporalAccessor;
+import java.util.HashMap;
+import java.util.Map;
+
+import static apoc.util.Util.getFormat;
+
+public class DateParseUtil {
+
+    public static Map<Class<? extends TemporalAccessor>, MethodHandle> parseDateMap = new HashMap<>();
```
Comment (jexp) [inline src/main/java/apoc/util/DateParseUtil.java]: and probably using a CHM for concurrent access.

File: src/main/java/apoc/util/DateParseUtil.java (hunk 6 above)
Comment (jexp) [inline src/main/java/apoc/util/DateParseUtil.java]: computeIfAbsent would make most sense.

File: src/main/java/apoc/util/DateParseUtil.java (hunk 7 above)
Comment (jexp) [inline src/main/java/apoc/util/DateParseUtil.java]: and probably using a CHM for concurrent access.

File: src/main/java/apoc/util/DateParseUtil.java (hunk 6 above)
Comment (conker84) [reply src/main/java/apoc/util/DateParseUtil.java]: ok

File: src/main/java/apoc/util/DateParseUtil.java (hunk 7 above)
Comment (conker84) [reply src/main/java/apoc/util/DateParseUtil.java]: ok

**Pull request merged**